source/common/network/utility.h

```diff
@@ -290,31 +291,49 @@ class Utility {
 
   /**
    * Send a packet via given UDP socket with specific source address.
-   * @param socket is the UDP socket used to send.
+   * @param handle is the UDP socket used to send.
    * @param slices points to the buffers containing the packet.
    * @param num_slices is the number of buffers.
    * @param local_ip is the source address to be used to send.
    * @param peer_address is the destination address to send to.
    */
-  static Api::IoCallUint64Result writeToSocket(Network::Socket& socket, Buffer::RawSlice* slices,
+  static Api::IoCallUint64Result writeToSocket(IoHandle& handle, Buffer::RawSlice* slices,
                                                uint64_t num_slices, const Address::Ip* local_ip,
                                                const Address::Instance& peer_address);
+  static Api::IoCallUint64Result writeToSocket(IoHandle& handle, const Buffer::Instance& buffer,
+                                               const Address::Ip* local_ip,
+                                               const Address::Instance& peer_address);
 
   /**
-   * Read a packet from given UDP socket and pass the packet to given
-   * UdpPacketProcessor.
-   * @param socket is the UDP socket to read from.
+   * Read a packet from a given UDP socket and pass the packet to given UdpPacketProcessor.
+   * @param handle is the UDP socket to read from.
+   * @param local_address is the socket's local address used to populate port.
    * @param udp_packet_processor is the callback to receive the packet.
    * @param receive_time is the timestamp passed to udp_packet_processor for the
    * receive time of the packet.
    * @param packets_dropped is the output parameter for number of packets dropped in kernel. If the
    * caller is not interested in it, nullptr can be passed in.
    */
-  static Api::IoCallUint64Result readFromSocket(Network::Socket& socket,
+  static Api::IoCallUint64Result readFromSocket(IoHandle& handle,
+                                                const Address::Instance& local_address,
                                                 UdpPacketProcessor& udp_packet_processor,
                                                 MonotonicTime receive_time,
                                                 uint32_t* packets_dropped);
 
+  /**
+   * Read all available packets from a given UDP socket and pass the packet to a given
+   * UdpPacketProcessor.
+   * @param handle is the UDP socket to read from.
+   * @param local_address is the socket's local address used to populate port.
+   * @param udp_packet_processor is the callback to receive the packets.
+   * @param time_source is the time source used to generate the time stamp of the received packets.
+   * @param packets_dropped is the output parameter for number of packets dropped in kernel.
+   */
+  static Api::IoCallUint64Result
+  readAllDatagramsFromSocket(IoHandle& handle, const Address::Instance& local_address,
```

Comment:
Will rename/add TODO/comment. I'm not sure about TCP as I don't think TCP will ever use recvmsg and this would be embedded inside the transport socket, but yes possibly. Will add a TODO/comment.